src/core/unit.c

```diff
@@ -2228,6 +2228,7 @@ void unit_start_on_failure(
                         logged = true;
                 }
 
+                other->triggered_by = u;
```

Comment:
Mmh I wonder if this could become invalid by the time the other side processes it? For example, a transient unit is immediately garbage collected when it exits successfully, but what happens in that case with OnSuccess?
Perhaps it's better to take a copy here of the couple of fields you are actually passing over?

Reply:
Try to implement that in the test - use systemd-run --property OnSuccess= and see what happens

Reply:
The new atom UNIT_ATOM_BACK_REFERENCE_IMPLIED should prevent the unit from being collected, though let me test your suggestions. Even still, the coupling here isn't as tight as I would like.

Reply:
Seems to work as expected:

```
root@localhost:~# systemd-run --unit=testfail --property=OnFailure=failer-handler.service /bin/bash -c "exit 1;"
Running as unit: testfail.service
root@localhost:~# journalctl -b -u testfail.service
-- Journal begins at Fri 2021-12-03 16:12:24 GMT, ends at Fri 2021-12-03 16:23:4
0 GMT. --
Dec 03 16:23:40 localhost systemd[1]: Started /bin/bash -c exit 1;.
Dec 03 16:23:40 localhost systemd[1]: testfail.service: 
Main process exited, code=exited, status=1/FAILURE
Dec 03 16:23:40 localhost systemd[1]: testfa
il.service: Failed with result 'exit-code'.                                     
Dec 03 16:23:40 localhost systemd[1]: testfail.service: Triggering OnFailure= de
pendencies.
root@localhost:~# 
```

The handler service runs fine:

```
root@localhost:~# systemctl cat failer-handler.service --no-pager
# /usr/lib/systemd/system/failer-handler.service
[Unit]
Description=Handle failures

[Service]
ExecStart=/bin/bash -c "echo $MONITOR_SERVICE_RESULT, $MONITOR_EXIT_CODE, $MONITOR_EXIT_STATUS, $MONITOR_INVOCATION_ID, $MONITOR_UNIT"

root@localhost:~# systemctl status failer-handler.service --no-pager -l
○ failer-handler.service - Handle failures
     Loaded: loaded (/usr/lib/systemd/system/failer-handler.service; static)
     Active: inactive (dead) since Fri 2021-12-03 16:23:40 GMT; 1min 53s ago
    Process: 515 ExecStart=/bin/bash -c echo $MONITOR_SERVICE_RESULT, $MONITOR_EXIT_CODE, $MONITOR_EXIT_STATUS, $MONITOR_INVOCATION_ID, $MONITOR_UNIT (code=exited, status=0/SUCCESS)
   Main PID: 515 (code=exited, status=0/SUCCESS)
        CPU: 1ms

Dec 03 16:23:40 localhost systemd[1]: Started Handle failures.
Dec 03 16:23:40 localhost bash[515]: exit-code, exited, 1, b4c76f2959aa4dcbaad008a4a27ec163, testfail.service
Dec 03 16:23:40 localhost systemd[1]: failer-handler.service: Deactivated successfully.
root@localhost:~# 
```

So the UNIT_ATOM_BACK_REFERENCE_IMPLIED atom is keeping the triggering service "testfail.service" pinned correctly.

I'll beef up the tests with something like this.

Reply:
check OnSuccess - when it fails it's not collected

Reply:
```
root@localhost:~# systemd-run --unit=testsuccess --property=OnSuccess=failer-handler.service /bin/bash -c "exit 0;"
Running as unit: testsuccess.service
root@localhost:~# 

root@localhost:~# systemctl status failer-handler.service --no-pager -l○ failer-handler.service - Handle failures
     Loaded: loaded (/usr/lib/systemd/system/failer-handler.service; static)
     Active: inactive (dead) since Fri 2021-12-03 16:30:45 GMT; 55s ago
    Process: 529 ExecStart=/bin/bash -c echo $MONITOR_SERVICE_RESULT, $MONITOR_EXIT_CODE, $MONITOR_EXIT_STATUS, $MONITOR_INVOCATION_ID, $MONITOR_UNIT (code=exited, status=0/SUCCESS)
   Main PID: 529 (code=exited, status=0/SUCCESS)
        CPU: 1ms

Dec 03 16:30:45 localhost systemd[1]: Started Handle failures.
Dec 03 16:30:45 localhost bash[529]: success, exited, 0, ccf6f89ec7f04899a58b1e690c80539e, testsuccess.service
Dec 03 16:30:45 localhost systemd[1]: failer-handler.service: Deactivated successfully.
root@localhost:~# 
```

```
root@localhost:~# systemctl status testsuccess --no-pager -l
○ testsuccess.service - /bin/bash -c exit 0;
     Loaded: loaded (/run/systemd/transient/testsuccess.service; transient)
  Transient: yes
     Active: inactive (dead) since Fri 2021-12-03 16:30:45 GMT; 2min 37s ago
    Process: 528 ExecStart=/bin/bash -c exit 0; (code=exited, status=0/SUCCESS)
   Main PID: 528 (code=exited, status=0/SUCCESS)
        CPU: 2ms

Dec 03 16:30:45 localhost systemd[1]: Started /bin/bash -c exit 0;.
Dec 03 16:30:45 localhost systemd[1]: testsuccess.service: Deactivated successfully.
Dec 03 16:30:45 localhost systemd[1]: testsuccess.service: Triggering OnSuccess= dependencies.
root@localhost:~# 
```

Reply:
ok, let's add it to the tests

Reply:
Done, thanks.  @bluca PTAL.

Reply:
looks good - maybe expand the commet above, saying it will be used for OnSuccess/OnFailure

Reply:
Done.

Reply:
One more thought: what about multiple activation in a rapid succession? eg, foo.service and bar.service both fail at the same time, the activation jobs get queued together, but won't `other->triggered_by` store whichever comes last?

Reply:
Yep, I think you are right. I'll need to have a think about this now.

Reply:
Store it in the `Job` struct instead of the unit?

Reply:
Could maybe do that? I need to check if I can access the job structure when we come to Exec the dependency.

Reply:
probably needs to be passed down when the spawn stack when the job is activated, but should be possible, and should be safe

Reply:
I had a go at doing that, seems to work fine.  PTAL!

Reply:
hmm, CI failures - I need to take a look.

Reply:
Fixed.

Reply:
one more thought (sorry): jobs can be merged too when it's the same job type and unit, what happens in that case? Should this be saved as a set that can be merged together when jobs merge, and the env vars deliver a space separate list instead of a single id/time/etc?

Reply:
We could make jobs that have different "triggered_by" units not mergeable, this might be a nice simplification?

Reply:
Actually, it's not that simple. I'll take a deeper look at this today.

Reply:
But then the second one would either be a no-op or fail because it's already running (not sure, would need to check), end result would be losing the information, so back to square one, no? I think the only alternative is explicitly documenting and enforcing that this is supported only for templated units, as then each activation is unique by definition, but it would mean it can never be extended, as the env var protocol cannot be changed once released

Reply:
I ended up adding a list of units to the Job struct, so this allows us to correctly merge lists of units when jobs to start the same unit are merged.

I still need to update the docs if you are happy with this approach & the format of the variable etc.

@bluca PTAL!